rust/otap-dataflow/crates/telemetry/src/instrument.rs

```diff
@@ -0,0 +1,500 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+//! The different types of instrument used to record the metric.
+//!
+//! These instruments are designed to be used in thread-per-core scenarios.
+//!
+//! ToDo Finish the implementation of UpDownCounter and Gauge (clean_values and needs_flush need some massage).
+//! ToDo Add histogram support
+
+use std::fmt::Debug;
+use std::ops::{AddAssign, SubAssign};
+
+/// A value that can only go up or be reset to 0, used for counts.
+#[repr(transparent)]
+#[derive(Default, Clone, Copy)]
+pub struct Counter<T>(T);
+
+/// A value that can go up and down, used for sizes or amount of items in a queue.
+#[repr(transparent)]
+#[derive(Default, Clone, Copy)]
+pub struct UpDownCounter<T>(T);
+
+/// A value that can arbitrarily go up and down, used for temperature or current memory usage.
+#[repr(transparent)]
+#[derive(Default, Clone, Copy)]
+pub struct Gauge<T>(T);
+
+// Counter implementation.
+// =======================
+
+impl<T: Copy + Default> Counter<T> {
+    /// Creates a new counter initialized with the provided value.
+    #[inline]
+    pub const fn new(v: T) -> Self {
+        Self(v)
+    }
+
+    /// Reset the counter to 0.
+    #[inline]
+    pub fn reset(&mut self) {
+        self.0 = T::default();
+    }
+
+    /// Returns the current value of the counter.
+    #[inline]
+    pub fn get(&self) -> T {
+        self.0
+    }
+}
+
+impl Debug for Counter<u64> {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Counter").field("value", &self.0).finish()
+    }
+}
+
+impl From<u64> for Counter<u64> {
+    fn from(value: u64) -> Self {
+        Counter(value)
+    }
+}
+
+impl AddAssign<u64> for Counter<u64> {
+    fn add_assign(&mut self, rhs: u64) {
+        #[cfg(feature = "unchecked-arithmetic")]
+        {
+            // SAFETY: Counter values are expected to be well-behaved in telemetry scenarios.
+            // Wrapping behavior is acceptable for performance-critical metric collection.
+            self.0 = self.0.wrapping_add(rhs);
+        }
+        #[cfg(not(feature = "unchecked-arithmetic"))]
+        {
+            self.0 += rhs;
+        }
+    }
+}
+
+impl Counter<u64> {
+    /// Increments the counter by 1.
+    #[inline]
+    pub fn inc(&mut self) {
+        #[cfg(feature = "unchecked-arithmetic")]
+        {
+            // SAFETY: Incrementing by 1 is safe for wrapping arithmetic in telemetry contexts
+            self.0 = self.0.wrapping_add(1);
```

Comment:
`wrapping_add` isn't an unsafe operation. Did you mean to use `unchecked_add` for this feature?

Reply:
> But at the same time, it's safer than += 1, which will panic.

@lquerel I'm not sure if you missed this https://github.com/open-telemetry/otel-arrow/pull/946#discussion_r2302769261, but `+=` will not panic in Release mode, so it's actually safe.